Error behaviors (including `onError: "NULL"`)

benjie opened the pull request:
This is a rewrite of:

- #1153

Service capabilities has been split into:

- #1208 

Depending on the order that these are merged, whichever is merged later should add the relevant service capabilities for error behaviors (each are optional):

- `graphql.onError`: indicates that this service accepts and honors the `onError` request property
- ~~`graphql.defaultErrorBehavior`: one of `PROPAGATE`, `NULL` or `ABORT`. Must be set unless the default error behavior is `PROPAGATE`. Allows creating a new service which uses a different error behavior from the outset. Changing this once a service is in production would be a breaking change.~~
  - Per the working group on 2026-08-06 we will forever have PROPAGATE as the default error behavior, so this capability has been removed. But worry not! The proposal is that the `onError` value to use can be incorporated into the [trusted documents](https://benjie.dev/graphql/trusted-documents) to save you precious network bytes.

---

Replaces:
- #1050 
- #1145
- #1153

GraphQL.js implementation:
- https://github.com/graphql/graphql-js/pull/4364

---

[Please see this (slightly outdated) 60 second video on the motivation for this PR](https://www.youtube.com/shorts/gYnVaZz-19A) (the last few seconds of the video also covers "transitional non-null" which is a [separate concern](https://github.com/graphql/graphql-spec/pull/1065)).

As agreed at the nullability working group, disabling error propagation is the future of error handling in GraphQL. Error propagation causes a number of issues, but chief among them are:

1. It destroys useful data in the response.
2. It makes it [unsafe to store resulting data in normalized stores](https://github.com/graphql/nullability-wg/discussions/23).

Clients such as Relay do not want error propagation to be a thing.

This has traditionally resulted in schema design best practices advising using nullable in positions where errors were expected, even if `null` was never a semantically valid value for that position. And since errors can happen everywhere, this has lead to an explosion of nullability and significant pain on the client side with developers having to do seemingly unnecessary null checks in loads of positions, or worse - unsafely bypassing the type safety.

The reason that GraphQL does error propagation is to keep it's "not null" promise in the event that an error occurs (and is replaced with `null` due to the way GraphQL responses are structured and limitations in JSON) in a non-nullable position.

It doesn't take much code on the client to prevent the client reading a `null` that relates to an error, [graphql-toe](https://www.npmjs.com/package/graphql-toe) can be used with almost any JavaScript or TypeScript based GraphQL client (not Relay, but it has `@throwOnFieldError` that you can use instead) and achieves this [in 512 bytes gzipped](https://bundlephobia.com/package/graphql-toe@1.0.0-rc.0) - and that's with a focus on performance rather than bundle size.

This PR allows the client to take responsibility for error handling by specifying `onError: "NULL"` as part of the GraphQL request, and thereby turns off error propagation behavior. This is also set as the recommended default for future schemas.

With clients responsible for error handling, we no longer need to factor the possibility of whether something can error or not into its nullability, meaning we can use the not-null `!` to indicate all the positions in the schema for which `null` is not a semantically valid value - i.e. the underlying resource will never be a legitimate `null`.

The end result:
- true nullability indicated in schema - no more thinking about where errors are likely
- fewer null checks on clients
- clients can leverage their native error handling capabilities such as try/catch or `<ErrorBoundary />`
- safe to store errored responses into normalized stores

---

I've also included `onError: "HALT"` in this proposal. We've discovered that there's a small but significant class of clients out there, mostly ad-hoc scripts, that throw away the entire response when any error occurs. By codifying this into the spec we make it easier to implement these clients, and we allow the server to halt processing the rest of the request unnecessarily.

As noted by @revangen in [this comment](https://github.com/graphql/graphql-spec/pull/1153#issuecomment-2810602063):

> > I've also included onError: ~~"ABORT"~~ "HALT" in this proposal.
>
> Appreciate this being included. For Shopify's public [Admin GraphQL API](https://shopify.dev/docs/api/admin-graphql), we have a mix of scenarios that result in a partial success response and only error response. Having been around for 8+ years, we are reluctant at times to change its behaviour to favour partial responses as we don't control majority of clients. Providing clients a way to specify the server's behaviour provides a migration path should clients care about partial responses.

benjie committed:
Detail onError request parameter

benjie committed:
Detail introspection changes

benjie committed:
Define the directive

benjie commented:
I have applied the RFC1 label to this as inherited from the PR it replaces:

- https://github.com/graphql/graphql-spec/pull/1153

benjie commented:
A full, CI-passing, implementation of this is available in GraphQL.js now; check it out: `graphql@canary-pr-4364`

martinbonnin commented on spec/Section 4 -- Introspection.md:
```diff
@@ -164,6 +165,12 @@ enum __TypeKind {
   NON_NULL
 }
 
+enum __ErrorBehavior {
+  NO_PROPAGATE
+  PROPAGATE
+  ABORT
```
Moving my [previous comment](https://github.com/graphql/graphql-spec/pull/1153#discussion_r2016918168) here, love this proposal overall but I would love it even more with `NULL`

```graphql
enum __ErrorBehavior {
  NULL
  PROPAGATE
  ABORT
}
```

benjie replied:
To summarize my thoughts on it from the previous thread: `NULL` is what I originally wrote, I agree it looks and initially feels nicer; but ultimately I discounted it for a number of reasons:

### Communication issues

If you tell someone on the phone to set `onError: NULL`, they're very likely to instead set `onError: null`.

### Ambiguity issues

`onError: null` and `onError: NULL` are very similar, but the first means `PROPAGATE` and the latter means `NO_PROPAGATE`. This is likely to be the source of many user errors and much frustration (both for newbies, and for the people supporting them including library authors).

### Errors _already_ become null in the spec:

> [an error] raised [...] is handled as though the response position [...] resolved to `null`
> -- https://spec.graphql.org/draft/#sel-FANVNJCAACGW-qR

`onError: NULL` doesn't implicitly seem to mean something different from this.

### Please suggest other alternatives!

I'm very open to changing the word `NO_PROPAGATE` to something else, I really don't like `NO_PROPAGATE`, not least because I always want to spell it propOgate rather than propAgate. But I think changing it to `NULL` would be a mistake.

I've gone through about 30 alternatives when brainstorming it, and none are sufficiently clear whilst also being shorter than `NO_PROPAGATE`. My favourite alternatives are things like `LOCAL`/`LOCALIZE`/`ISOLATE`/`CONTAIN`, but I'm not sure they convey the meaning quite as well as `NO_PROPAGATE` does.

benjie replied:
Ooo... Maybe `LOCAL_NULL` actually is quite compelling. Not thought of that one before.

fotoetienne commented on spec/Section 6 -- Execution.md:
```diff
@@ -15,13 +17,23 @@ A GraphQL service generates a response from a request via execution.
   being executed. Conceptually, an initial value represents the "universe" of
   data available via a GraphQL Service. It is common for a GraphQL Service to
   always use the same initial value for every request.
+- {onError} (optional): The _error behavior_ to apply to the request; see
```
What about onError (or behavior) as a query directive rather than a new field on the request? Adding a field to the request payload might require changes to lots of GraphQL middleware

martinbonnin replied:
This is where we initially started with `@disableErrorPropagation`. It's very handy that no middleware is impacted at all but @benjie [argumented that it should be a request parameter](https://github.com/graphql/nullability-wg/discussions/86).

The argument that won me is that the directive would most likely be automatically added by the Apollo/Relay/... compilers. Then it means you're modifying the user query and if the user types the same query in GraphiQL it won't behave the same, which is surprising.

fotoetienne replied:
That's a good point that you might want to be able to add it without modifying the query document

benjie replied:
Also keep in mind that it being a GraphQL request parameter does not necessarily mean it’s an HTTP request parameter; for example you could use a different endpoint in HTTP to set the param: `/graphql_nobubblesplz` ;)

fotoetienne commented on spec/Section 3 -- Type System.md:
````diff
@@ -2167,3 +2167,27 @@ to the relevant IETF specification.
+### @behavior
+
+```graphql
+directive @behavior(onError: __ErrorBehavior! = PROPAGATE) on SCHEMA
````
Thoughts on `directive @onError(behavior: __ErrorBehavior!) on SCHEMA` to mirror the naming of the client `onError` field?

benjie replied:
Here’s [why I think it should be a request parameter and not a directive](https://github.com/graphql/nullability-wg/discussions/86). (Further, I don’t think we should allow both because that opens up questions around what happens if the client sets one and the user sets the other: who wins?)

fotoetienne replied:
What about defining the onError behavior only in introspection and not in the SDL? We can then define an abstract "configuration" block that informs execution behavior. Implementations can then choose to make this be defined as a schema directive or as a separate configuration file.

benjie committed:
ABORT -> HALT

GraphQLConf requires speakers to be familiar with "inclusive naming",
and the inclusive naming guides encourage the avoidance of the word
"abort" where possible:

https://inclusivenaming.org/word-lists/tier-1/abort/

benjie committed:
Start speccing out the capabilities system

benjie committed:
Add a number of basic capabilities

benjie committed:
Move default error behavior to the service

benjie committed:
Rework capabilities

benjie committed:
Use a definition

benjie committed:
Reorder

benjie committed:
Reword

benjie committed:
Editorial

benjie committed:
More editorial

benjie committed:
More editorial

benjie commented:
This PR description is currently out of date; I've reworked the capabilities infrastructure based on Lee's feedback, see the changes for details.

martinbonnin commented:
@benjie @leebyron given the previous discussions on feature discovery have been going on for years, I'm doubtful we can land `__Capability` soon... Yet `onError` is crucially needed to move the nullability proposals, is there any chance we could use a new introspection field (`__Schema.defaultErrorBehaviour`) instead? Or should we expedite `__Capability` in a separate proposal first? (I'm down with that FWIW)

benjie commented:
@martinbonnin You don’t like this proposal?

martinbonnin commented on spec/Section 4 -- Introspection.md:
```diff
@@ -500,3 +510,76 @@ Fields\:
   used repeatedly at a single location.
+
+### The \_\_Service Type
+
+The `__Service` type is returned from the `__service` meta-field and provides
+information about the GraphQL service, most notably about its capabilities. This
+type was added after the original release of GraphQL, so older schemas may not
+support it.
```
Do we really need the `__service` indirection vs having `__capabilities`? 
My mental model is that the endpoint is the service. It has a `__schema` and `__capabilities` are at the same level (except they can't be expressed in terms of type system).

martinbonnin replied:
Yikes, I'm having an existential crisis now...

Can clients know whether `__service` is supported? As a meta field, it's not returned in introspection, right? So it's too late now, we have to retrofit everything inside `__schema`???

martinbonnin commented on spec/Section 4 -- Introspection.md:
```diff
@@ -500,3 +510,76 @@ Fields\:
+
+A capability identifier is a string value composed of two or more segments
+separated by a period (`.`). Each segment must begin with an ASCII letter, and
+must contain only ASCII letters, digits and hyphens. These constraints are
+inspired by reverse domain name notation to encourage global uniqueness and
+collision-resistance. Unlike the domain name system, capability identifiers are
```
I like this!

martinbonnin commented on spec/Section 4 -- Introspection.md:
```diff
@@ -500,3 +510,76 @@ Fields\:
+
+This version of the specification defines the following capabilities:
+
+- {"org.graphql.defaultErrorBehavior"} - indicates the _default error behavior_
+  of the service via the {value}. If not present, must be assumed to be
+  {"PROPAGATE"}.
+- {"org.graphql.onError"} - indicates that the service allows the client to
+  specify {onError} in a request to indicate the _error behavior_ the service
+  should use for the request. {value} is {null}.
```
What about this?

```suggestion
- {"org.graphql.defaultErrorBehavior"} - indicates the _default error behavior_
  of the service via the {value}. 
  If present, the service must be assumed to support the {onError} parameter 
  to indicate the _error behavior_ on requests.
  If not present, must be assumed to be {"PROPAGATE"}.
```

Less invalid cases possible.

benjie replied:
This requires services to support the `onError` request parameter forever, I don't think we should do that. If a service wants to start up that only supports the `NO_PROPAGATE` error behavior and doesn't want to allow clients to influence that, I think it should be allowed to.

martinbonnin replied:
SGTM 👍

alex-reilly-dd commented on spec/Section 4 -- Introspection.md:
```diff
@@ -500,3 +510,76 @@ Fields\:
+
+**Capability identifier**
+
+A capability identifier is a string value composed of two or more segments
```
Clarify that this is a suggestion

benjie replied:
Governed by the "preserve option value" guiding principle:

> It's hard to know what the future brings; whenever possible, decisions should be made that allow for more options in the future. Sometimes this is unintuitive: spec rules often begin more strict than necessary with a future option to loosen when motivated by a real use case.

I think we should keep this as a requirement rather than a suggestion/recommendation. It's quite simple to comply with, and we can always loosen it up in future if need be.

E.g. maybe we might want to add an SDL syntax for this in future, it would be a shame to have to wrap all the keys in strings:

```
service {
  capabilities {
    org.graphql.onError
    org.graphql.defailtErrorBehavior: "PROPAGATE"
    org.graphql.operationDescription
    org.graphql.ws.subscriptions.endpoint: "wss://example.com/graphql"
    org.graphql.ws.queries
    org.graphql.ws.mutations
    com.graphile.env: "Production"
    com.graphile.hash: "sdf9009s8dg09sdf809"
    org.graphql.rfc.ccn: "v1"
    org.graphql.rfc.incrementalDelivery: "June2023"
  }
}
```

That said, I'm not too bothered and happy to remove it if there's a good reason.

benjie committed:
Update spec/Section 4 -- Introspection.md

benjie commented:
It could be split, but I would do the capabilities feature first and then onError so we don't need a period of time where you send onError and hope - better to be concrete from the beginning IMO as client capabilities may depend on it being honoured. It makes sense to ship capabilities alongside something that the spec itself needs it for in my opinion - could be error behaviors, could be fragment arguments, but I think it should be something.

michaelstaib commented:
We have discussed the service capabilities proposal last week in the Composite Schema Working Group, and I wanted to share our perspective on this.

We'd strongly advocate for having service capabilities represented in the SDL. This is essential for the schema composition tooling as only SDL exposes type system directives and the introspection format does not carry these yet. Also most tooling for composing schemas does not dynamically introspect a source schemas but rather works without needing to start a GraphQL server against schema files.

Concretely, we are discussing for quite some time how we can inspect capabilities like batching support. Depending on if a server supports batching or not the schema composition would allow for features like cross-service data requirements.

There are other things that we think would be great to inspect in that process like the schema name and other metadata.

We did consider introducing our own capabilities directives to the specification, that would attach to the schema definition node. But to be honest this always felt like duck taping these things into GraphQL.

We agree that we should keep it simple with a string-base value for service capabilities. This allows us to start using it immediately without complicating the spec too much. It also leaves room for future enhancements, such as structured metadata or enums, as the ecosystem evolves.

We also think this should be decoupled from the error proposal as this would solve so many other concerns in GraphQL.

benjie committed:
Merge branch 'main' into error-behavior2

benjie committed:
Change NO_PROPAGATE to NULL

benjie commented on spec/Section 4 -- Introspection.md:
```diff
@@ -218,6 +219,15 @@ enum __DirectiveLocation {
 }
+
+type __Service {
+  capabilities: [__Capability!]!
```
I've thought about this quite a bit, and I think capabilities is better - it's more indicative of intent/purpose.

The aim is that clients can use the "capabilities" of the service to inform their own behaviors in an automated way, such that end users can just point to a GraphQL endpoint and the client will take care of the rest - they don't need to configure it extensively (and the configuration can be automatically updated on each release without frontend engineers needing to do anything).

For example:

- does it have the capability of handling fragment arguments? If not, I must transpile them.
- does it have the capability of handling document directives? If not, I must find them invalid
- does it have the capability of customizing error propagation behavior? If not, I must assume that it will propagate errors (and should not pass `onError`)
- does it support subscriptions over websockets? SSE? If so, what are the endpoints?
- does it support automated document persistence? If so, which protocol does it use and what are the relevant configuration parameters?
- does it support notification of new schema releases? If so, at which endpoint can I subscribe to these events? (Especially useful for GraphiQL)
- does it support incremental delivery? Which version?
- does it support {RFC feature X}?
- etc

The capabilities system _can_ be used for metadata, but that's not its purpose; and calling it metadata significantly changes what is being proposed (and also encourages it to be used for feeding a huge amount more "metadata" into the system). Capabilities inform automated client decisions and actions; metadata is ancillary and often intended for human consumption.

I will try and ensure this is correctly reflected in the spec text.

benjie committed:
Overhaul and introduce SDL syntax for service capabilities

benjie committed:
Update syntax

benjie committed:
Simplify

benjie committed:
Update Name syntax

benjie committed:
Clarification

benjie committed:
Minor tweaks

benjie committed:
Clarify the interaction of onError:HALT and subscriptions

benjie committed:
Fix grammar

benjie committed:
Spelling

benjie committed:
Fix incorrect prefix

leebyron commented on spec/Appendix B -- Grammar Summary.md:
```diff
@@ -47,7 +47,7 @@ Punctuator :: one of ! $ & ( ) ... : = @ [ ] { | }
 Name ::
 
-- NameStart NameContinue\* [lookahead != NameContinue]
+- NameStart NameContinue\* [lookahead != {NameContinue, `.`}]
```
don't think we need this,

the reason for this quirk in the grammar is to remove ambiguity about greediness, e.g.

```
ThisIsAName
```

Where `ThisIs` would be a valid match for `NameStart NameContinue*`, but isn't what we want since we want it to greedily consume the rest of the following NameContinue, hence we say `[lookahead != NameContinue]` to make it clear that it's not valid to stop here and it must greedily continue before yielding a Name token

leebyron commented on spec/Appendix B -- Grammar Summary.md:
```diff
@@ -413,3 +419,19 @@ TypeSystemDirectiveLocation : one of
+
+- ServiceCapabilities
+
+ServiceCapabilities: capabilities { ServiceCapability+ }
```
`*` instead of `+` and drop the group-less one?

I know historically we've done this "at least one required" grammatically instead of with validation + to do this extension-without-add pattern, but I think we've learned thats quite annoying in IDEs and tools. So I think we should explore dropping it here

```suggestion
ServiceDefinition :

- Description? service { ServiceAttribute* }

ServiceAttribute :

- ServiceCapabilities

ServiceCapabilities: capabilities { ServiceCapability* }
```

thoughts?

benjie committed:
Remove unnecessary negative lookahead

benjie committed:
Lee says empty braces are okay here!

Co-authored-by: Lee Byron <lee@leebyron.com>

benjie committed:
Merge branch 'main' into error-behavior2

benjie committed:
Sync syntax with Lee's suggestions

benjie commented on spec/Section 4 -- Introspection.md:
````diff
@@ -85,13 +85,14 @@ operation.
 ```graphql
 __schema: __Schema!
 __type(name: String!): __Type
+__service: __Service!
````
Discussed at WG, conclusion was:

- we do not need to include this in the spec
- we should never add anything that requires runtime introspection of service capabilities to the spec - build time should always suffice
- **if someone is disabling introspection in production, there's a very good chance that they're doing the wrong thing and should be using Trusted Documents instead**

martinbonnin commented:
@benjie any thoughts about breaking this in 2 different proposals? 

1. `onError`
2. `serviceCapabilities`

I can see a bunch of cases where teams could start using `1.` without necessarily requiring `2.` (because they would communicate the default out of band). And that would make the adoption path for frameworks more incremental.

Thoughts?

Edit: I've just noticed https://github.com/graphql/graphql-spec/pull/1208 👍 , let's do the same for `onError`?

bbarry commented on spec/Section 6 -- Execution.md:
```diff
@@ -600,13 +611,26 @@ section.
+- {"NULL"}: The _response position_ must be set to {null}, even if such position
+  is indicated by the schema to be non-nullable. (The client is responsible for
+  interpreting this {null} in conjunction with the {"errors"} list to
+  distinguish error results from intentional {null} values.)
```
In the case of mutation resolution, does `"NULL"` mean execution continues?

eg if I have a mutation:

```graphql
mutation {
  doThing1
  doThing2
  doThing3
}
```

and all fields were not nullable, which of these responses would be correct in the presence of an error on resolution of `doThing2`:

A: response fields deviate from graphql request ask

```json
{
  "data": {
    "doThing1": true,
    "doThing2": null
    /* doThing3 never resolved */
  },
  "errors": [
    { /* error from doThing2 */ }
  ]
}
```

B: synthetic field production - propagate null to down-operation resolution

```json
{
  "data": {
    "doThing1": true,
    "doThing2": null,
    "doThing3": null /* never resolved but server injects null because field is requested */
  },
  "errors": [
    { /* error path doThing2 */ }
  ]
}
````

C: continue resolving

```json
{
  "data": {
    "doThing1": true,
    "doThing2": null,
    "doThing3": null /* server application's responsibility for stopping execution after error */
  },
  "errors": [
    { /* error path doThing2 */ },
    { /* error path doThing3 thrown due to response errored flag set, does not indicate a problem with doThing3 */ }
  ]
}
```

or perhaps more confusingly

```json
{
  "data": {
    "doThing1": true,
    "doThing2": null,
    "doThing3": true
  },
  "errors": [
    { /* error from doThing2 */ }
  ]
}
```

benjie replied:
The final “perhaps more confusingly” is the correct behavior according to the current semantics. It’s an interesting question, because if the client changes the onError without the knowledge of the user, the resulting side-effects will differ. Option C is interesting, but also wrong in its own way. In my own schemas, root level fields (even mutations) are nullable so it wouldn’t make a difference to me, I wonder if making the fields non-nullable is done specifically to block follow-up mutations currently…? Though it does mean you wouldn’t see the result of previously completed mutations so it seems like a weird choice.

bbarry replied:
Hot Chocolate's mutation convention makes mutations not nullable and it is quite annoying to change: https://chillicream.com/docs/hotchocolate/v16/building-a-schema/mutations/#mutation-conventions and I think semantically it is correct (the mutation always has a result or there is an error). 

I agree the current wording suggests option C as well (both of the last 2 samples are option C). The difference between them is if the server implementer (or framework they use) flag that an error occurred via some mechanism such that if inside `ExecuteMutation()` the flag is set, every following `ExecuteSelectionSet()` immediately raises an error. 

I suppose my point is that either application authors (both clients and servers) will need to beware of of the possibility that this could happen (and frameworks could take an opinionated stance or provide a default) or the spec force A or B and clients (and potentially client libraries) would need to be able to handle such responses.  I don't think there is a satisfying answer here.

martinbonnin replied:
@bbarry can you ellaborate why C. is not satisfying? 

`onError: NULL` makes it possible to resolve more data, whether for queries or subscriptions. Forcing A. or B. would be very surprising to me. With C. the server can decide how it wants to handle things. If anything, we could recommend to have only a single root field in mutations. I think this was discussed at some point but can't find it anymore.

Edit: found it! It's "batched mutations" article: https://medium.com/@xuorig/graphql-mutation-design-batch-updates-ca2452f92833

bbarry replied:
I think it is a very hard problem and every solution I've seen so far is unsatisfying in some way. In each case there is room for interpretation and miscommunication between client and server. If clients could just send valid data and servers could avoid flaky systems the world would be a much simpler place.

What I find unsatisfying about C is that a client a client sending such a mutation operation could intend that the 3 mutation fields are causality related. If the second doesn't happen and the 3rd does, there could be a logic bug in the client.  If the server author chooses to avoid this potential by raising down-field errors the server is adding unnecessary response content and falsely stating there is a problem with a field when in fact the problem is outside the field.

benjie committed:
Merge branch 'main' into error-behavior2

benjie committed:
Remove service capabilities; now in #1208

benjie committed:
Add pathNonNull so clients can reproduce propagation locally

benjie committed:
Add pathNonNull to examples

benjie committed:
Better wording

benjie committed:
No capabilities

leebyron commented on spec/Section 6 -- Execution.md:
```diff
@@ -15,13 +15,21 @@ A GraphQL service generates a response from a request via execution.
 
+Note: Previous versions of this specification did not define the {onError}
+request attribute. Clients should not include {onError} in the request unless it
+is know that the service supports this property.
```
is known

leebyron commented on spec/Section 6 -- Execution.md:
```diff
@@ -15,13 +15,21 @@ A GraphQL service generates a response from a request via execution.
 formatted according to the Response section below.
 
+Note: Previous versions of this specification did not define the {onError}
+request attribute. Clients should not include {onError} in the request unless it
```
should only include {onError} ... if it is known that the service

benjie committed:
Remove the concept of a default error behavior

benjie committed:
Less awkward wording, per Lee
